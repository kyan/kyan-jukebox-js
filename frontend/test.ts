#!/usr/bin/env bun
/**
 * Simplified Bun Test Runner with Jest Compatibility
 *
 * Uses Bun's Jest compatibility mode to run existing Jest-style tests
 * without requiring complex migration of test files.
 */

import { spawnSync } from 'child_process'
import { existsSync } from 'fs'

interface TestOptions {
  watch?: boolean
  coverage?: boolean
  verbose?: boolean
  silent?: boolean
  bail?: boolean
  watchAll?: boolean
}

class BunJestRunner {
  private options: TestOptions = {}

  constructor(args: string[] = []) {
    this.parseArgs(args)
  }

  private parseArgs(args: string[]): void {
    for (let i = 0; i < args.length; i++) {
      const arg = args[i]

      switch (arg) {
        case '--watch':
        case '-w':
          this.options.watch = true
          break
        case '--watchAll':
          this.options.watchAll = true
          break
        case '--watchAll=false':
          this.options.watchAll = false
          this.options.watch = false
          break
        case '--coverage':
          this.options.coverage = true
          break
        case '--verbose':
          this.options.verbose = true
          break
        case '--silent':
          this.options.silent = true
          break
        case '--bail':
          this.options.bail = true
          break
        case '--forceExit':
        case '--detectOpenHandles':
        case '--maxWorkers=1':
          // Legacy Jest options - ignore as Bun handles these automatically
          break
      }
    }

    // Default watch behavior
    if (
      !process.env.CI &&
      this.options.watchAll === undefined &&
      this.options.watch === undefined
    ) {
      this.options.watch = true
    }
  }

  private setupEnvironment(): void {
    // Set test environment variables
    process.env.NODE_ENV = 'test'
    process.env.BABEL_ENV = 'test'
    process.env.PUBLIC_URL = ''
  }

  private async runPreflightChecks(): Promise<void> {
    if (this.options.silent) return

    console.log('üß™ Running preflight checks...\n')

    // Type checking
    if (existsSync('./tsconfig.json')) {
      console.log('üìù Type checking...')
      try {
        const typeCheck = spawnSync('bun', ['run', 'type-check'], {
          stdio: 'inherit'
        })

        if (typeCheck.status !== 0) {
          console.error('‚ùå Type check failed')
          process.exit(1)
        } else {
          console.log('‚úÖ Type check passed')
        }
      } catch (error) {
        console.warn('‚ö†Ô∏è Type checking skipped:', error.message)
      }
    }

    // Linting
    console.log('üîç Linting...')
    try {
      const lint = spawnSync('bun', ['run', 'lint'], {
        stdio: 'inherit'
      })

      if (lint.status !== 0) {
        console.error('‚ùå Linting failed')
        process.exit(1)
      } else {
        console.log('‚úÖ Linting passed')
      }
    } catch (error) {
      console.warn('‚ö†Ô∏è Linting skipped:', error.message)
    }

    // Format check
    console.log('üíÖ Format checking...')
    try {
      const format = spawnSync('bun', ['run', 'format:check'], {
        stdio: 'inherit'
      })

      if (format.status !== 0) {
        console.error('‚ùå Format check failed')
        console.log('üí° Run: bun run format')
        process.exit(1)
      } else {
        console.log('‚úÖ Format check passed')
      }
    } catch (error) {
      console.warn('‚ö†Ô∏è Format checking skipped:', error.message)
    }

    console.log('')
  }

  private buildJestCommand(): { command: string; args: string[] } {
    const args: string[] = []

    // Test pattern - Jest will find spec files automatically
    args.push('--testMatch', '<rootDir>/src/**/*.{spec,test}.{js,jsx,ts,tsx}')

    // Watch mode
    if (this.options.watchAll) {
      args.push('--watchAll')
    } else if (this.options.watch) {
      args.push('--watch')
    } else {
      args.push('--watchAll=false')
    }

    // Coverage
    if (this.options.coverage) {
      args.push('--coverage')
    }

    // Verbose mode
    if (this.options.verbose) {
      args.push('--verbose')
    }

    // Silent mode
    if (this.options.silent) {
      args.push('--silent')
    }

    // Bail on first failure
    if (this.options.bail) {
      args.push('--bail')
    }

    // Force exit and handle detection for CI
    if (process.env.CI || !this.options.watch) {
      args.push('--forceExit', '--detectOpenHandles', '--maxWorkers=1')
    }

    return { command: 'bunx', args: ['jest', ...args] }
  }

  async run(): Promise<void> {
    if (!this.options.silent) {
      console.log('üèÉ‚Äç‚ôÇÔ∏è Bun Jest Test Runner\n')
    }

    // Setup test environment
    this.setupEnvironment()

    // Run preflight checks (validation)
    await this.runPreflightChecks()

    // Build and run Jest command through Bun
    const { command, args } = this.buildJestCommand()

    if (this.options.verbose && !this.options.silent) {
      console.log('üìã Command:', [command, ...args].join(' '))
    }

    if (!this.options.silent) {
      console.log('üß™ Running tests...\n')
    }

    try {
      const result = spawnSync(command, args, {
        stdio: 'inherit',
        cwd: process.cwd(),
        env: {
          ...process.env,
          // Ensure Bun uses Jest compatibility mode
          FORCE_COLOR: '1'
        }
      })

      if (result.status !== 0) {
        if (!this.options.silent) {
          console.error('\n‚ùå Tests failed')
        }
        process.exit(result.status || 1)
      } else {
        if (!this.options.silent) {
          console.log('\n‚úÖ All tests passed!')
        }
      }
    } catch (error) {
      console.error('‚ùå Test runner failed:', error.message)

      if (!this.options.silent) {
        console.log('\nüí° Troubleshooting:')
        console.log('   ‚Ä¢ Ensure Bun v1.0+ is installed: bun --version')
        console.log('   ‚Ä¢ Try: bun install (to ensure dependencies are correct)')
        console.log('   ‚Ä¢ Check Jest configuration in package.json')
      }

      process.exit(1)
    }
  }
}

// CLI execution
if (import.meta.main) {
  const args = process.argv.slice(2)

  if (args.includes('--help') || args.includes('-h')) {
    console.log(`
Bun Jest Test Runner

Usage:
  bun test.ts [options]

Options:
  --watch, -w          Watch files for changes
  --watchAll           Watch all files
  --watchAll=false     Disable watch mode (for CI)
  --coverage           Generate coverage report
  --verbose            Verbose output
  --silent             Suppress output except errors
  --bail               Stop on first test failure

Examples:
  bun test.ts                           # Run all tests (watch mode in dev)
  bun test.ts --watchAll=false         # CI mode - run once
  bun test.ts --coverage               # With coverage report
  bun test.ts --silent --watchAll=false # Silent CI mode
`)
    process.exit(0)
  }

  const runner = new BunJestRunner(args)
  runner.run().catch(error => {
    console.error('Test runner failed:', error)
    process.exit(1)
  })
}
