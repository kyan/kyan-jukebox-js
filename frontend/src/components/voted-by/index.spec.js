import { describe, it, expect, beforeEach } from 'bun:test'
import React from 'react'
import { render } from '@testing-library/react'
import VotedBy from './index'

describe('VotedBy', () => {
  beforeEach(() => {
    // Bun mocks are cleared automatically between tests
  })

  describe('render', () => {
    const mockVotes = [
      {
        at: '2020-01-27T20:20:37.904Z',
        vote: 80,
        user: {
          _id: '1',
          fullname: 'John Doe',
          picture: 'https://example.com/john.jpg'
        }
      },
      {
        at: '2020-01-27T20:20:50.355Z',
        vote: 60,
        user: {
          _id: '2',
          fullname: 'Jane Smith',
          picture: 'https://example.com/jane.jpg'
        }
      }
    ]

    it('renders with votes when show is true', () => {
      const { container } = render(<VotedBy show={true} votes={mockVotes} total={70} />)
      expect(container.firstChild).toBeInTheDocument()
    })

    it('renders nothing when show is false', () => {
      const { container } = render(<VotedBy show={false} votes={mockVotes} total={70} />)
      expect(container.firstChild).toBeNull()
    })

    it('renders just vote label when no votes provided', () => {
      const { container } = render(<VotedBy show={true} votes={[]} total={50} />)
      expect(container.firstChild).toBeInTheDocument()
    })

    it('renders with positive vote total', () => {
      const { container } = render(<VotedBy show={true} votes={[]} total={80} />)
      expect(container.firstChild).toBeInTheDocument()
    })

    it('renders with negative vote total', () => {
      const { container } = render(<VotedBy show={true} votes={[]} total={20} />)
      expect(container.firstChild).toBeInTheDocument()
    })

    it('renders with ribbon prop', () => {
      const { container } = render(
        <VotedBy show={true} votes={mockVotes} total={70} ribbon={true} />
      )
      expect(container.firstChild).toBeInTheDocument()
    })

    it('renders with custom size', () => {
      const { container } = render(<VotedBy show={true} votes={[]} total={50} size='small' />)
      expect(container.firstChild).toBeInTheDocument()
    })

    it('handles votes with missing user data', () => {
      const votesWithMissingUser = [
        {
          at: '2020-01-27T20:20:37.904Z',
          vote: 80,
          user: null
        }
      ]
      const { container } = render(<VotedBy show={true} votes={votesWithMissingUser} total={80} />)
      expect(container.firstChild).toBeInTheDocument()
    })
  })
})
