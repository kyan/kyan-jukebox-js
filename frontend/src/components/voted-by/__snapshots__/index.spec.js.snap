// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VotedBy when no user data does nothing 1`] = `
<Popup
  content={
    <List>
      <ListItem>
        <ListContent>
          <ListDescription>
            <strong>
              19 Dec 2019 @ 1:11 pm
            </strong>
             - 
            User unknown
             
            <Label
              circular={true}
              size="mini"
            >
              3
            </Label>
          </ListDescription>
        </ListContent>
      </ListItem>
    </List>
  }
  disabled={false}
  eventsEnabled={true}
  on={
    Array [
      "click",
      "hover",
    ]
  }
  pinned={false}
  popperModifiers={Array []}
  position="top left"
  trigger={
    <Label
      basic={true}
      className="track-label vote-track-label"
      color="grey"
      content={-4}
      icon="thumbs down"
      ribbon={null}
      size="tiny"
    />
  }
  wide="very"
/>
`;

exports[`VotedBy when no votes are provided just shows a label with +5 vote 1`] = `
<Label
  basic={false}
  className="track-label vote-track-label"
  color="green"
  content={5}
  icon="thumbs up"
  ribbon={null}
  size="tiny"
/>
`;

exports[`VotedBy when no votes are provided just shows a label with -3 vote 1`] = `
<Label
  basic={true}
  className="track-label vote-track-label"
  color="grey"
  content={-3}
  icon="thumbs down"
  ribbon={null}
  size="tiny"
/>
`;

exports[`VotedBy when no votes are provided just shows a label with 0 vote 1`] = `
<Label
  basic={true}
  className="track-label vote-track-label"
  color="grey"
  content={0}
  icon="thumbs up"
  ribbon={null}
  size="tiny"
/>
`;

exports[`VotedBy when no votes are provided just shows a label with 0 vote 2`] = `
<Label
  basic={false}
  className="track-label vote-track-label"
  color="red"
  content={-5}
  icon="thumbs down"
  ribbon={null}
  size="tiny"
/>
`;

exports[`VotedBy when show is false shows nothing 1`] = `""`;

exports[`VotedBy when votes information is provided and a ribbon prop is set displays the correct information 1`] = `
<Popup
  content={
    <List>
      <ListItem>
        <Image
          as="img"
          avatar={true}
          className="voted_by_avatar_image"
          src="link/to/image"
          ui={true}
        />
        <ListContent>
          <ListDescription>
            <strong>
              17 Dec 2019 @ 1:11 pm
            </strong>
             - 
            Big Rainbowhead
             
            <Label
              circular={true}
              size="mini"
            >
              -3
            </Label>
          </ListDescription>
        </ListContent>
      </ListItem>
      <ListItem>
        <Image
          as="img"
          avatar={true}
          className="voted_by_avatar_image"
          src="link/to/duncan/image"
          ui={true}
        />
        <ListContent>
          <ListDescription>
            <strong>
              19 Dec 2019 @ 1:11 pm
            </strong>
             - 
            Duncan
             
            <Label
              circular={true}
              size="mini"
            >
              3
            </Label>
          </ListDescription>
        </ListContent>
      </ListItem>
    </List>
  }
  disabled={false}
  eventsEnabled={true}
  on={
    Array [
      "click",
      "hover",
    ]
  }
  pinned={false}
  popperModifiers={Array []}
  position="top left"
  trigger={
    <Label
      basic={true}
      className="track-label vote-track-label"
      color="grey"
      content={-4}
      icon="thumbs down"
      ribbon="right"
      size="tiny"
    />
  }
  wide="very"
/>
`;

exports[`VotedBy when votes information is provided and no ribbon prop displays the correct information 1`] = `
<Popup
  content={
    <List>
      <ListItem>
        <Image
          as="img"
          avatar={true}
          className="voted_by_avatar_image"
          src="link/to/image"
          ui={true}
        />
        <ListContent>
          <ListDescription>
            <strong>
              17 Dec 2019 @ 1:11 pm
            </strong>
             - 
            Big Rainbowhead
             
            <Label
              circular={true}
              size="mini"
            >
              -3
            </Label>
          </ListDescription>
        </ListContent>
      </ListItem>
      <ListItem>
        <Image
          as="img"
          avatar={true}
          className="voted_by_avatar_image"
          src="link/to/duncan/image"
          ui={true}
        />
        <ListContent>
          <ListDescription>
            <strong>
              19 Dec 2019 @ 1:11 pm
            </strong>
             - 
            Duncan
             
            <Label
              circular={true}
              size="mini"
            >
              3
            </Label>
          </ListDescription>
        </ListContent>
      </ListItem>
    </List>
  }
  disabled={false}
  eventsEnabled={true}
  on={
    Array [
      "click",
      "hover",
    ]
  }
  pinned={false}
  popperModifiers={Array []}
  position="top left"
  trigger={
    <Label
      basic={true}
      className="track-label vote-track-label"
      color="grey"
      content={-3}
      icon="thumbs down"
      ribbon={null}
      size="tiny"
    />
  }
  wide="very"
/>
`;
